### ISSUE TITLE
Workaround for plugin/loop on Kubernetes not working

### ISSUE THREAD

**Issue by jforman** [coredns/coredns#2790]
Hi all,

I'm trying to diagnose a failing coredns deploy to my bare-metal kubernetes setup that fails on detecting a loop. I have followed the directions at https://coredns.io/plugins/loop/#troubleshooting, and specified my resolv conf in the kubelet config file, but something must be not quite right.

Any help is greatly appreciated.

Stats:

- Ubuntu 18.04 LTS libvirt/kvm VM
- Kubernetes 1.14.0
- Coredns 1.5.0

CoreDNS pod log:
```
.:53
2019-04-17T00:18:44.485Z [INFO] plugin/reload: Running configuration MD5 = fe280f2b52a55a24ba57f2b4328c2777
2019-04-17T00:18:44.485Z [INFO] CoreDNS-1.5.0
2019-04-17T00:18:44.485Z [INFO] linux/amd64, go1.12.2, e3f9a80
CoreDNS-1.5.0
linux/amd64, go1.12.2, e3f9a80
2019-04-17T00:18:44.486Z [FATAL] plugin/loop: Loop (127.0.0.1:34614 -> :53) detected for zone ".", see https://coredns.io/plugins/loop#troubleshooting. Query: "HINFO 6074102630077018525.5798689655824409358.
```

Pod description:
```
Name:               coredns-85767f454b-b5fcq
Namespace:          kube-system
Priority:           0
PriorityClassName:  <none>
Node:               devb-worker0.domain.obfuscated.net/10.10.2.133
Start Time:         Tue, 16 Apr 2019 20:17:17 -0400
Labels:             k8s-app=kube-dns
                    pod-template-hash=85767f454b
Annotations:        container.apparmor.security.beta.kubernetes.io/coredns: runtime/default
                    kubernetes.io/psp: 999-privileged
                    seccomp.security.alpha.kubernetes.io/pod: docker/default
Status:             Running
IP:                 10.244.3.10
Controlled By:      ReplicaSet/coredns-85767f454b
Containers:
  coredns:
    Container ID:  containerd://1e051fb5ae5d59f185996c4791d454d3b6359220cbda8f2925812aead8bb00c0
    Image:         coredns/coredns:1.5.0
    Image ID:      docker.io/coredns/coredns@sha256:e83beb5e43f8513fa735e77ffc5859640baea30a882a11cc75c4c3244a737d3c
    Ports:         53/UDP, 53/TCP, 9153/TCP
    Host Ports:    0/UDP, 0/TCP, 0/TCP
    Args:
      -conf
      /etc/coredns/Corefile
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Tue, 16 Apr 2019 20:20:11 -0400
      Finished:     Tue, 16 Apr 2019 20:20:11 -0400
    Ready:          False
    Restart Count:  5
    Limits:
      memory:  170Mi
    Requests:
      cpu:        100m
      memory:     70Mi
    Liveness:     http-get http://:8080/health delay=60s timeout=5s period=10s #success=1 #failure=5
    Environment:  <none>
    Mounts:
      /etc/coredns from config-volume (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from coredns-token-fr9ct (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  config-volume:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      coredns
    Optional:  false
  coredns-token-fr9ct:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  coredns-token-fr9ct
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  beta.kubernetes.io/os=linux
Tolerations:     CriticalAddonsOnly
                 node-role.kubernetes.io/master:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                    From                                                   Message
  ----     ------     ----                   ----                                                   -------
  Normal   Scheduled  4m44s                  default-scheduler                                      Successfully assigned kube-system/coredns-85767f454b-b5fcq to devb-worker0.domain.obfuscated.net
  Normal   Pulling    4m42s                  kubelet, devb-worker0.domain.obfuscated.net  Pulling image "coredns/coredns:1.5.0"
  Normal   Pulled     4m37s                  kubelet, devb-worker0.domain.obfuscated.net  Successfully pulled image "coredns/coredns:1.5.0"
  Normal   Created    3m17s (x5 over 4m37s)  kubelet, devb-worker0.domain.obfuscated.net  Created container coredns
  Normal   Started    3m17s (x5 over 4m37s)  kubelet, devb-worker0.domain.obfuscated.net  Started container coredns
  Normal   Pulled     3m17s (x4 over 4m37s)  kubelet, devb-worker0.domain.obfuscated.net  Container image "coredns/coredns:1.5.0" already present on machine
  Warning  BackOff    3m16s (x9 over 4m36s)  kubelet, devb-worker0.domain.obfuscated.net  Back-off restarting failed container
```

# kubelet-config.yaml
```
......
ResolverConfig: "/run/systemd/resolve/resolv.conf"
```

# /run/systemd/resolve/resolv.conf 
```
nameserver 10.10.2.1
```

# Corefile
```
   .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          upstream
          fallthrough in-addr.arpa ip6.arpa
        }
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```

If I change /etc/resolv.conf to /run/systemd/resolv/resolv.conf in the Corefile, I get the following error:
```
plugin/forward: not an IP address or file: "/run/systemd/resolve/resolv.conf"
```

**Comment by chrisohaver** [comment]
You won't want to change the path to resolv.conf in the Pod itself.  That will always be at `/etc/resolv.conf`.  Kublet essentially takes the `resolv.conf` file you give it, and puts that in `/etc` for all Pods using the "default" Pod DNS policy (e.g. the CoreDNS Pod).

On your nodes:
* What is in`/run/systemd/resolve/resolv.conf` 
* And what is in `/etc/resolv.conf`

After configuring kubelet to use `/run/systemd/resolve/resolv.conf` did you restart the CoreDNS Pods?  I dont know if kubelet does this automatically.  CoreDNS would need to restart to see the change (the `reload` plugin only detects changes to the Corefile, not to files it references).

**Comment by jforman** [comment]
> You won't want to change the path to resolv.conf in the Pod itself. That will always be at `/etc/resolv.conf`. Kublet essentially takes the `resolv.conf` file you give it, and puts that in `/etc` for all Pods using the "default" Pod DNS policy (e.g. the CoreDNS Pod).
> 
> On your nodes:
> 
> * What is in`/run/systemd/resolve/resolv.conf`

```shell
ubuntu@devb-worker1:~$ cat /run/systemd/resolve/resolv.conf 
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# This is a dynamic resolv.conf file for connecting local clients directly to
# all known uplink DNS servers. This file lists all configured search domains.
#
# Third party programs must not access this file directly, but only through the
# symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a different way,
# replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 10.10.2.1
```


> * And what is in `/etc/resolv.conf`

```shelll
ubuntu@devb-worker1:~$ cat /etc/resolv.conf 
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# This is a dynamic resolv.conf file for connecting local clients to the
# internal DNS stub resolver of systemd-resolved. This file lists all
# configured search domains.
#
# Run "resolvectl status" to see details about the uplink DNS servers
# currently in use.
#
# Third party programs must not access this file directly, but only through the
# symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a different way,
# replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 127.0.0.53
options edns0
```

> 
> After configuring kubelet to use `/run/systemd/resolve/resolv.conf` did you restart the CoreDNS Pods?

Yes, I've created and deleted the deployment several times, to no avail.

> I dont know if kubelet does this automatically. CoreDNS would need to restart to see the change (the `reload` plugin only detects changes to the Corefile, not to files it references).

**Comment by chrisohaver** [comment]
@jforman, couple of things to try to narrow down the issue...

1. Can you do look ups from Pods directly against your upstream servers? e.g. `dig @10.10.2.1 example.com`
2. If you remove the `loop` plugin, can you successfully do upstream queries from Pods? ... e.g. `dig example.com` ?  (trying to determine if the loop plugin is signaling a false positive)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Workaround fo..." ]
> (trying to determine if the loop plugin is signaling a false positive)

interested to learn how that would happen...?

/Miek

--
Miek Gieben

**Comment by jforman** [comment]
> @jforman, couple of things to try to narrow down the issue...
> 
> 1. Can you do look ups from Pods directly against your upstream servers? e.g. `dig @10.10.2.1 example.com`

Note: No CoreDNS installed on any node in my k8s cluster. 

root@devb-worker2:/usr/bin #nslookup google.com 10.10.2.1
Server:    10.10.2.1
Address 1: 10.10.2.1 gate1.obfuscated.domain.net

Name:      google.com
Address 1: 2607:f8b0:400d:c02::64 qm-in-x64.1e100.net
Address 2: 173.194.66.100 qo-in-f100.1e100.net
Address 3: 173.194.66.102 qo-in-f102.1e100.net
Address 4: 173.194.66.113 qo-in-f113.1e100.net
Address 5: 173.194.66.101 qo-in-f101.1e100.net
Address 6: 173.194.66.138 qo-in-f138.1e100.net
Address 7: 173.194.66.139 qo-in-f139.1e100.net

> 2. If you remove the `loop` plugin, can you successfully do upstream queries from Pods? ... e.g. `dig example.com` ?  (trying to determine if the loop plugin is signaling a false positive)

Sigh, I might have had a misconfiguration somewhere in my config (it had been at least since coredns 1.2 since I last tried coredns) or my cluster, because with the loop plugin enabled, now it works.

root@devb-worker2:~ #nslookup google.com 10.200.0.11
Server:    10.200.0.11
Address 1: 10.200.0.11

Name:      google.com
Address 1: 2607:f8b0:400d:c0a::64 ql-in-x64.1e100.net
Address 2: 173.194.205.138 qm-in-f138.1e100.net
Address 3: 173.194.205.113 qm-in-f113.1e100.net
Address 4: 173.194.205.100 qm-in-f100.1e100.net
Address 5: 173.194.205.139 qm-in-f139.1e100.net
Address 6: 173.194.205.101 qm-in-f101.1e100.net
Address 7: 173.194.205.102 qm-in-f102.1e100.net

(10.200.0.11 being the service address for core-dns)

```
% kubectl logs -l k8s-app=kube-dns -n kube-system             
.:53
2019-04-24T00:42:21.779Z [INFO] plugin/reload: Running configuration MD5 = fe280f2b52a55a24ba57f2b4328c2777
2019-04-24T00:42:21.779Z [INFO] CoreDNS-1.5.0
2019-04-24T00:42:21.779Z [INFO] linux/amd64, go1.12.2, e3f9a80
CoreDNS-1.5.0
linux/amd64, go1.12.2, e3f9a80
.:53
2019-04-24T00:42:22.863Z [INFO] plugin/reload: Running configuration MD5 = fe280f2b52a55a24ba57f2b4328c2777
2019-04-24T00:42:22.863Z [INFO] CoreDNS-1.5.0
2019-04-24T00:42:22.863Z [INFO] linux/amd64, go1.12.2, e3f9a80
CoreDNS-1.5.0
linux/amd64, go1.12.2, e3f9a80
```

For those who come finding this issue in the future, my Corefile is as follows:
```
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          upstream
          fallthrough in-addr.arpa ip6.arpa
        }
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```

Sorry for the noise, and I appreciate the patient debugging!

**Comment by jforman** [comment]
Interesting wrinkle, since I was still curious why it seemed to work some ways, and not others.

From my sample size of 1, it looks like it all depends on how your kubelet is configured.

When configuring the Kubelet's resolv.conf the deprecated way, CoreDNS works.

What is the deprecated way? Specifying the resolv.conf via the --resolv-conf flag (--resolv-conf=/run/systemd/resolve/resolv.conf) and NOT specifying the ResolverConfig in the Kubelet YAML config (ResolverConfig: "/run/systemd/resolve/resolv.conf"). 

If I specified the resolv.conf via the YAML and not the kubelet flag, things broke with the following log output:
```
.:53
2019-04-25T23:30:57.067Z [INFO] plugin/reload: Running configuration MD5 = f73ead3fd97ca11d31973ad6d1565a14
2019-04-25T23:30:57.067Z [INFO] CoreDNS-1.5.0
2019-04-25T23:30:57.067Z [INFO] linux/amd64, go1.12.2, e3f9a80
CoreDNS-1.5.0
linux/amd64, go1.12.2, e3f9a80
2019-04-25T23:30:57.068Z [FATAL] plugin/loop: Loop (127.0.0.1:49602 -> :53) detected for zone ".", see https://coredns.io/plugins/loop#troubleshooting. Query: "HINFO 2829883173337881696.442540980926758253."
```

Kubelet log complaints when using the --resolv-conf flag:
```
Apr 25 23:41:37 devb-worker2 kubelet[2737]: Flag --resolv-conf has been deprecated, This parameter should be set via the config file specified by 
the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
```

To circle back on @chrisohaver 's earlier debugging questions, with a running CoreDNS with the kubelet configured in the deprecated way:

```
/ $ nslookup www.google.com 10.122.0.10
Server:    10.122.0.10
Address 1: 10.122.0.10 kube-dns.kube-system.svc.cluster.local

Name:      www.google.com
Address 1: 209.85.201.106 qu-in-f106.1e100.net
Address 2: 209.85.201.105 qu-in-f105.1e100.net
Address 3: 209.85.201.104 qu-in-f104.1e100.net
Address 4: 209.85.201.147 qu-in-f147.1e100.net
Address 5: 209.85.201.99 qu-in-f99.1e100.net
Address 6: 209.85.201.103 qu-in-f103.1e100.net
Address 7: 2607:f8b0:400d:c07::69 qb-in-x69.1e100.net
/ $ nslookup www.google.com 10.10.2.1
Server:    10.10.2.1
Address 1: 10.10.2.1 gate1.obfuscated.domain.net

Name:      www.google.com
Address 1: 209.85.201.147 qu-in-f147.1e100.net
Address 2: 209.85.201.104 qu-in-f104.1e100.net
Address 3: 209.85.201.106 qu-in-f106.1e100.net
Address 4: 209.85.201.99 qu-in-f99.1e100.net
Address 5: 209.85.201.105 qu-in-f105.1e100.net
Address 6: 209.85.201.103 qu-in-f103.1e100.net
Address 7: 2607:f8b0:400d:c07::69 qb-in-x69.1e100.net
```

So to conclude, this sounds like something odd/different in our the kubelet presents DNS information depending on how it's configured. For now, I'm going to back to configuring my kubelet in the deprecated way.

If folks on CoreDNS have knowledge of the kubelet or connections with folks who do, please feel free to share my issue!

**Comment by chrisohaver** [comment]
~Thanks for digging into it. Do you have a link/reference that describes the command flag as deprecated?~

Uh, yeah its right there in the output. :)

**Comment by chrisohaver** [comment]
@jforman, Actually - ~this suddenly seems familiar...  I recall a peculiarity with the option name in the kubelet yaml.  IIRC, I think the name may actually "ResolvConf" or "Resolve-Conf" ... not "ResolverConfig" (as documented)...~   ... Never mind, see next comment.

**Comment by chrisohaver** [comment]
Correction ... I think that the aforementioned option name discrepancy was a problem in the minikube `extra-config` option for setting options in kubelet, not directly in kubelet ...

**Comment by chrisohaver** [comment]
>So to conclude, this sounds like something odd/different in our the kubelet presents DNS information depending on how it's configured. For now, I'm going to back to configuring my kubelet in the deprecated way.
>
> If folks on CoreDNS have knowledge of the kubelet or connections with folks who do, please feel free to share my issue!

IMO, this looks like a problem in kubelet.  Assuming that the config yaml is actually being used by kubelet in your case, it seems that kubelet fails to acknowledge the `ResolverConfig` option.  
There is no direct interaction between this option and CoreDNS.  CoreDNS simply uses whatever is its Pod's `/etc/resolv.conf`.  When Kubelet creates a Pod, it (is supposed to) set up `/etc/resolv.conf` according to the option.

**Comment by jforman** [comment]
@chrisohaver Your mention of the 'ResolverConfig' option made me dig back into the kubelet code.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L473.

`ResolverConfig` is the TYPE, `resolvConf` is the variable name referenced in the json/yaml configuration. The former is what I had used as the variable name prior, but it is the latter that is correct.

Once my YAML looked like
```
resolvConf: "/run/systemd/resolve/resolv.conf"
```

all was working, loop plugin as well.

Turns out it wasn't an issue with the kubelet, but merely my configuration of it. Hopefully this debugging-the-obvious is helpful for others. 

Cheers,
